Docs and improvements for Orderbook (#164)
Origin-Byte/nft-protocol: docs/ORDERBOOK.md

# Orderbook

Orderbook implementation where bids are fungible tokens and asks are NFTs.
A bid is a request to buy one NFT from a specific collection.
An ask is one NFT with a min price condition.

`Ask` is an object which is associated with a single NFT.
When `Ask` is created, we transfer the ownership of the NFT to this new object.
To be more precise, we transfer the `safe::TransferCap`.

One can:

- create a new orderbook between a given collection and a BID token (witness
  pattern protected)
- set publicly accessible actions to be witness protected
- open a new BID
- cancel an existing BID they own
- offer an NFT if a collection matches OB collection
- cancel an existing NFT offer
- instantly buy a specific NFT

# Intermediary state

> By working with `Safe`, clients must know _up front_ which exact `Safe`
> instance to provide into the entry methods for transferring an NFT.
> However, OB concept supports frequent trades with `create_ask` and `create_bid`
> endpoints.
> In frequent trades, the NFT changes fast and therefore the `Safe` instance
> cannot reliably be known in advance.

Reviewer: ```suggestion
When trading frequently, the `Safe` in which an NFT lives often cannot be known
in advance. This becomes a problem as clients must know _up front_ which exact
`Safe` instance to provide into the `create_ask` and `create_bid` orderbook
endpoints.

```

Author: Let me reword it, but 

> This becomes a problem as clients must know _up front_ which exact `Safe` instance to provide into the `create_ask` and `create_bid` orderbook

Is not correct. They must know if only for transfer. In fact, it is the case that for those two endpoints they don't need to know it by design.

Reviewer: > This becomes a problem as clients would have to know up front which exact Safe instance to provide into the create_ask and create_bid orderbook

Perhaps would be better. What was unclear to me was whether you were describing the hypothetical problem situation or the solution, so I suggested to rewrite describing the problem situation which would then be followed by the solution.

the seller's `Safe` objects as arguments.
The `TradeIntermediate` is a shared object so that both parties can actually
drive the trade to completion.

# Commission

When a bid or an ask is created via a wallet or a marketplace, the client can
use `create_bid_with_commission` or `create_ask_with_commission` endpoints.
These endpoints have two additional arguments: `beneficiary` address
and `commission_ft` amount.

Creating a bid with a commission amount will take this amount from the buyer's
wallet and lock them along with the funds they are bidding with.
Once the bid is matched, then the locked commission funds are instantly
transferred to the beneficiary address.

Creating an ask with a commission amount will take the amount from the seller's
reward for the NFT.
For example, if an NFT is sold for 10 SUI and commission was 2 SUI, then the
NFT seller receives 8 SUI and the beneficiary receives 2 SUI.
Once the ask is matched, then the commission funds are _eventually_ transferred
to the beneficiary address.
See the [documentation](#intermediary-state) for the intermediary state above.
The commission is paid to the beneficiary only after the intermediary state
have been resolved.

> To receive a commission on an instant NFT buy (when buying a specific NFT)

Reviewer: Unclear what you mean

Author: reworded

To create a new instance of an orderbook which trades given collection for given
fungible token call the following endpoint:

```move
create<C, FT>()
```

To create a new bid, the client provides the `Safe` into which they wish to
receive an NFT.
They provide the price in the smallest unit of the fungible token.
This amount will be taken from the provided `Coin` wallet.

This endpoint will either store a new bid in the orderbook, or it will match
the bid with an existing ask thereby executing the trade.

In such a case, a new shared object [`TradeIntermediate`](#intermediary-state)
is created.

```move
create_bid<C, FT>(
  book: &mut Orderbook<C, FT>,
  buyer_safe: &mut Safe,
  price: u64,
  wallet: &mut Coin<FT>,
)
```

In addition to the above, the client can ask for a [commission](#commission)
when they create the bid on behalf of the signer.

```move
create_bid_with_commission<C, FT>(
  book,
  buyer_safe,
  price: u64,
  beneficiary: address,
  commission_ft: u64,
  wallet,
)
```

To cancel an existing bid, the client gives the price they sent as an input in
the previous endpoint.
If there are multiple bids the tx sender has created with the same price, then

> only one is cancelled.

Reviewer: ```suggestion
only the first order is cancelled.
```

the seller's `Safe` objects as arguments.
The `TradeIntermediate` is a shared object so that both parties can actually
drive the trade to completion.

# Commission

When a bid or an ask is created via a wallet or a marketplace, the client can
use `create_bid_with_commission` or `create_ask_with_commission` endpoints.
These endpoints have two additional arguments: `beneficiary` address
and `commission_ft` amount.

Creating a bid with a commission amount will take this amount from the buyer's
wallet and lock them along with the funds they are bidding with.
Once the bid is matched, then the locked commission funds are instantly
transferred to the beneficiary address.

Creating an ask with a commission amount will take the amount from the seller's
reward for the NFT.
For example, if an NFT is sold for 10 SUI and commission was 2 SUI, then the
NFT seller receives 8 SUI and the beneficiary receives 2 SUI.
Once the ask is matched, then the commission funds are _eventually_ transferred
to the beneficiary address.
See the [documentation](#intermediary-state) for the intermediary state above.
The commission is paid to the beneficiary only after the intermediary state
have been resolved.

> If a marketplace or a wallet want to receive a commission on a trade they

Reviewer: ```suggestion
If a marketplace or a wallet wants to receive a commission on a trade they
```
